File: built-in plugins dsp/mb-wt-massive.cpp
/**
 * MB Massive Bass
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Huge wavetable bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WT_MASSIVE_H
#define MB_WT_MASSIVE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWtMassive : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-wt-massive";
    static constexpr const char* PLUGIN_NAME    = "MB Massive Bass";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float position = 0.3f;  // range [0, 1]
    float intensity = 0.8f;  // range [0, 1]
    float volume = 0.85f;  // range [0, 1]
    };

    MbWtMassive() = default;
    ~MbWtMassive() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.position = std::clamp(params.position, 0f, 1f);
        params.intensity = std::clamp(params.intensity, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Massive Bass
        return input;
    }
};

#endif // MB_WT_MASSIVE_H

File: built-in plugins dsp/mb-wt-massive.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbWtMassivePlugin: PluginDefinition = {
  id: "mb-wt-massive",
  slug: "mb-wt-massive",
  name: "MB Massive Bass",
  category: "instrument",
  type: "wavetable",
  version: "1.0.0",
  description: "Huge wavetable bass",
  author: "Max Booster",
  grade: "A",
  oscillators: [{ type: "sawtooth", detune: 0, gain: 0.9 }],
  envelope: { attack: 0.001, decay: 0.3, sustain: 0.7, release: 0.2 },
  parameters: [
    {
      id: "position",
      name: "Wave Position",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "intensity",
      name: "Intensity",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.85,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { position: 0.3, intensity: 0.8, volume: 0.85 },
};

export default MbWtMassivePlugin;
